sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnarPartitionReaderWithPartitionValues.scala

```diff
@@ -114,4 +178,70 @@ object ColumnarPartitionReaderWithPartitionValues {
       }
     }
   }
+
+  // scalastyle:off line.size.limit
+  /**
+   * Splits the partition into smaller batches such that the batch size
+   * does not exceed cudfLimit for any column.
+   *
+   * Algorithm:
+   * - Calculate the max size of single value among partition columns.
+   * - Calculate the size of partition = numRows * maxSizeOfSingleValue.
+   * - If size of partition exceeds cudfLimit, split the partition into batches.
+   *    - Calculate newRows = cudfLimit / maxSizeOfSingleValue.
+   *    - Add newRows to resultBatches.
+   *    - Update remainingNumRows and remainingSizeOfPartition.
+   *    - Repeat until remainingSizeOfPartition is less than cudfLimit.
+   *  - Return the array of batch sizes in resultBatches.
+   *
+   * Example:
+   * Input:
+   *    numRows: 50
+   *    partValues: [ abc, abcdef ]
+   *    cudfLimit: 140 bytes
+   *
+   * Result:
+   *    [ 23, 23, 4 ]
+   *
+   * @return An array of batch sizes, such that the size of each batch is
+   *         below cudfLimit.
+   */
+  // scalastyle:on line.size.limit
+  private def splitPartitionIntoBatchesScalar(numRows: Int,
+      partValues: Array[Scalar], sparkTypes: Array[DataType]): Array[Int] = {
```

Comment:
Curious why this is working with the GPU type of `Scalar` here?  We can do the size calculations and thus determine the necessary splitting points from the CPU partition values (i.e.; InternalRow).  We only want to deal with `Scalar` just at the point when we're splatting those scalars into the partition columns, and at that point we should already know how many rows to use because we pre-calculated the split points.

Reply:
Makes sense. I will refactor it so that the split indices are calculated before. This was made here to minimize the difference between single partition values and multiple partition values.